keps/sig-windows/20190714-windows-csi-support.md

```diff
@@ -0,0 +1,462 @@
+2. Allow host path mounts with pathPrefix = `\\.\pipe\csi-proxy`. Restrict usage of this PSP to only SAs associated with the DaemonSets of CSI Node Plugins.
+Support will need to be implemented in AllowsHostVolumePath to handle Windows pipe paths.
+
+#### Mitigation using a webhook
+An admission webhook can be implemented and deployed in clusters with Windows nodes that will reject all containers that mount `\\.\pipe\csi-proxy` as a hostPath volume but does not have privileged flag set in the pod's securityContext specification. This allows the privileged setting to be used for Windows pods as an indication the container will perform privileged operations. Other cluster-wide policies (e.g. PSP) that act on the privileged setting in a container's securityContext can enforce the same for CSI Node plugins targeting Windows nodes. Note that this does not in any way change how the privileged setting is used today for Linux nodes. If in the future, full privileged container support is introduced in Windows (as in Linux today), functionality of existing CSI Node Plugin DaemonSets (targeting Windows) with the privileged flag set should not get negatively impacted as they will be launched as privileged containers.
+
+## Design Details
+
+### Test Plan
+
+Unit tests will be added to verify Windows related enhancements in existing Kubernetes components mentioned above.
+
+All E2E storage tests covering CSI plugins will be ported to Windows workloads and successfully executed with above enhancements in place along with csi-proxy.exe.
+
+### Graduation Criteria
+
+#### Alpha -> Beta Graduation
+
+- csi-proxy.exe supports v1beta1 version of the CSIProxyService API.
+- end-2-end tests in place with a CSI plugin that can support Windows containers and pass all existing CSI plugin test scenarios.
+
+#### Beta -> GA Graduation
+
+- In-tree storage plugins that implements support for Windows (AWS EBS, GCE PD, Azure File and Azure Disk as of today) can use csi-proxy.exe along with other enhancements listed above to successfully deploy CSI plugins on Windows nodes.
+- csi-proxy.exe supports v1 stable version of the CSIProxyService API.
+- Successful usage of csi-proxy.exe with support for v1 version of CSIProxyService API in Windows nodes by at least two storage vendors.
+
+### Upgrade / Downgrade Strategy
+
+In order to install a CSI Node Plugin or upgrade to a version of a CSI Node Plugin that uses an updated version of the CSIProxyService API not supported by the currently deployed version of csi-proxy.exe in the cluster, csi-proxy.exe will need to be upgraded first on all nodes of the cluster before deploying or upgrading the CSI Node Plugin. In case there is a very old CSI Node Plugin in the cluster that relies on a version of CSIProxyService API that is no longer supported by the new version of csi-proxy.exe, the previously installed version of csi-proxy.exe should not be uninstalled from the nodes. Such scenarios are expected to be an exception.
```

Comment:
Another strategy for the old apiversion scenario is to have csiproxy support multiple api versions for a few releases according to deprecation policy?

Reply:
I will call out a deprecation policy mirroring Kubernetes policy but again emphasize that goal should be to maintain compatibility as much as possible.